Sources/StreamChat/Workers/Background/NewChannelQueryUpdater.swift

```diff
@@ -60,58 +50,82 @@ final class NewChannelQueryUpdater: Worker {
+                    self?.linkChannelToExistedQueries($0.cid)
+                }
             } catch {
                 log.error("Error starting NewChannelQueryUpdater observer: \(error)")
             }
         }
     }
     
-    private func handle(changes: [ListChange<ChannelDTO>]) {
+    private func handle(changes: [ListChange<ChatChannel>]) {
         // Observe `ChannelDTO` insertions
         changes.forEach { change in
             switch change {
-            case let .insert(channelDTO, _):
-                let cid = channelDTO.channelId
-                
+            case let .insert(channel, _):
+                let cid = channel.cid
                 database.write {
                     let dto = $0.channel(cid: cid)
                     dto?.needsRefreshQueries = false
                 } completion: { _ in
-                    self.updateChannelListQuery(for: channelDTO)
+                    self.linkChannelToExistedQueries(cid)
                 }
 
             default: return
             }
         }
     }
     
-    private func updateChannelListQuery(for channelDTO: ChannelDTO) {
-        database.backgroundReadOnlyContext.perform { [weak self] in
-            guard let queries = self?.queries else { return }
-            
-            var updatedQueries: [ChannelListQuery] = []
-            
-            do {
-                updatedQueries = try queries.map {
-                    // Modify original query filter
-                    try $0.asChannelListQueryWithUpdatedFilter(filterToAdd: .equal("cid", to: channelDTO.channelId.rawValue))
+    private func linkChannelToExistedQueries(_ cid: ChannelId) {
+        fetchExistedQueries { [weak self] queries in
+            for query in queries {
+                self?.linkChannelToQueryIfNeeded(cid, query: query)
+            }
+        }
+    }
+    
+    private func linkChannelToQueryIfNeeded(_ cid: ChannelId, query: ChannelListQuery) {
+        var queryWithNewChannel = ChannelListQuery(
+            filter: .and([query.filter, .equal(.cid, to: cid)]),
+            pageSize: 1
+        )
+        queryWithNewChannel.options = []
+        
+        channelListUpdater.fetch(queryWithNewChannel) { [weak self] in
+            switch $0 {
+            case let .success(payload):
+                guard let channel = payload.channels.first(where: { $0.channel.cid == cid }) else {
+                    return
                 }
                 
-            } catch {
-                log.error("Internal error. Failed to update ChannelListQueries for the new channel: \(error)")
+                self?.save(channel, andLinkTo: query)
+            case let .failure(error):
+                log.error("Failed to check if query should include new channel: \(error)")
             }
+        }
+    }
+    
+    private func fetchExistedQueries(_ completion: @escaping ([ChannelListQuery]) -> Void) {
+        let context = database.backgroundReadOnlyContext
+        context.perform {
```

Comment:
What do you mean by "existed"?

Reply:
i meant `fetchLocalQueries`. will do the rename 👍